synapse/lib/crypto/coin.py

```diff
@@ -0,0 +1,150 @@
+import logging
+import binascii
+
+import cbor  # cardano
+import regex
+import bech32  # cardano
+import base58  # btc and cardano
+import _pysha3  # do not import the sha3 library directly.
+import bitcoin  # btc
+import bitcoin.bech32 as bitcoin_b32  # btc
+import synapse.vendor.cashaddress.convert as cashaddr_convert  # BCH support
+import synapse.vendor.xrpl.core.addresscodec as xrp_addresscodec  # xrp support
+import synapse.vendor.substrateinterface.utils.ss58 as substrate_ss58  # polkadot support
+
+import synapse.common as s_common
+
+logger = logging.getLogger(__name__)
+
+'''
+synapse.lib.crypto.coin contains functions for verifying whether or not a given
+regex match containing a valu is valid for a given type of coin.
+
+these functions are intended to be used with synapse.lib.scrape.
+'''
+
+
+def btc_bech32_check(match: regex.Match):
+    text = match.groupdict().get('valu')
+    prefix, _ = text.split('1', 1)
+    prefix = prefix.lower()
+    if prefix == 'bc':
+        bitcoin.SelectParams('mainnet')
+    elif prefix == 'tb':
+        bitcoin.SelectParams('testnet')
+    elif prefix == 'bcrt':
+        bitcoin.SelectParams('regtest')
+    else:  # pragma: no cover
+        raise ValueError(f'Unknown prefix {text}')
+    try:
+        _ = bitcoin_b32.CBech32Data(text)
+    except bitcoin_b32.Bech32Error:
+        return None
+    # The proper form of a bech32 address is lowercased. We do not want to verify
+    # a mixed case form, so lowercase it prior to returning.
+    return ('btc', text.lower())
+
+def btc_base58_check(match: regex.Match):
+    text = match.groupdict().get('valu')
+    try:
+        base58.b58decode_check(text)
+    except ValueError:
+        return None
+    return ('btc', text)
+
+def ether_eip55(body: str):
+    # From EIP-55 reference implementation
+    addr_byts = s_common.uhex(body)
+    hex_addr = addr_byts.hex()
+    checksummed_buffer = ""
+    hashed_address = _pysha3.keccak_256(hex_addr.encode()).hexdigest()
+
+    for nibble_index, character in enumerate(hex_addr):
```

Comment:
Is it just me, or is this whole loop just `hex_addr.upper()` ?

Reply:
It's selectively upper-casing alpha characters based on the hash. If the resulting mixed case string is not equal to the original mixed case string, that is a checksum mismatch.